File: 12/25/Min-Stack/Min-Stack.cpp
// CPP Solution 

```
   // O(N) Time +O(N)+O(N) Space complexity
   
   stack<int> sn;
stack<int> sm;
MinStack::MinStack() {
    // sn.clear();
    sn = stack<int>();
    sm = stack<int>();
}

void MinStack::push(int x) {
    sn.push(x);
    if(!sm.empty()){
        sm.push(min(sm.top(),x));
    }else{
        sm.push(x);
    }
}
void MinStack::pop() {
    if(!sn.empty()){
        sn.pop();
    sm.pop();
    }else{
        // Do Nothing
    }
    
}

int MinStack::top() {
    if(sn.empty()){
        return -1;
    }
    int ans= sn.top();
    return ans;
}

int MinStack::getMin() {
    if(!sm.empty()){
       return sm.top();   
    }
    return -1;
}



```

File: 12/25/Min-Stack/Min-Stack.java
// JAVA Solution

```
class Solution {
    Stack<Integer> s = new Stack<Integer>();
    int mValue = -1;
    public void push(int x) {
        if(s.isEmpty() == true){
            s.push(x);
            mValue = x;
            return ;
        }
        if(x<mValue){
            s.push(2*x-mValue);
            mValue= x;
        }else{
            s.push(x);
        }
    }
    public void pop() {
        if(s.isEmpty() == true){
            mValue=-1;
            return ;
        }
        int t = s.pop();
        if (t < mValue) {
            mValue = 2 * mValue - t;
        }
        if(s.isEmpty()== true){
            mValue=-1;
        }
        return ;
    }
    public int top() {
        if(s.isEmpty() == true){
            mValue =-1;
            return -1;
        }
        int top = s.peek();
        if(top<mValue){
            return mValue;
        }else{
            return top;
        }
    }
    public int getMin() {
        return mValue;
    }
}

```
